File: python/binary_search_tree/red-black-tree.py
# rotate a tree nd
# d = 0: right rotation
# d = 1: left rotation
def rotate(nd, d):
    c = nd[d]
    if d:
        e = c[1]
        nd[1] = c[0]
        c[0] = nd
    else:
        e = c[0]
        nd[0] = c[1]
        c[1] = nd

    r = c[3] = nd[3]
    nd[3] = r - (e[3] if e else 0) - 1
    return c

# insert a node with key = val into a tree t
def insert(t, val):
    x = t
    st = []; dr = []
    y = None
    while x:
        y = x
        st.append(x)
        if x[2] == val:
            return t
        d = (x[2] < val)
        dr.append(d)
        x = x[d]

    # [<left>, <right>, <key>, <count>, <color: [BLACK,RED]>]
    nd = [None, None, val, 1, 1]

    if not y:
        nd[4] = 0
        return nd

    y[d] = nd
    for x in st:
        x[3] += 1

    while 1:
        if not st:
            # if nd is the root node
            nd[4] = 0
            return nd

        # nd's parent
        y = st.pop(); dy = dr.pop()
        if not y[4]:
            # if nd's parent is black
            return t

        # nd's grandparent
        w = st.pop(); dw = dr.pop()
        # nd's uncle
        z = w[dw ^ 1]
        if z and z[4]:
            # the parent and the uncle is red
            y[4] = z[4] = 0
            w[4] = 1
            nd = w
            continue

        if dw != dy:
            w[dw] = rotate(y, dy)
            nd, y = y, nd

        rotate(w, dw)
        y[4] = 0
        w[4] = 1

        if st:
            st[-1][dr[-1]] = y
            return t
        return y

# delete a node nd
def __delete(st, dr, nd):
    x = nd
    if x[0]:
        st.append(x); dr.append(0)
        x = x[0]
        while x:
            st.append(x); dr.append(1)
            x = x[1]

        y = st.pop(); dr.pop()
        # copy a value only
        nd[2] = y[2]
    elif x[1]:
        st.append(x); dr.append(1)
        x = x[1]
        while x:
            st.append(x); dr.append(0)
            x = x[0]

        y = st.pop(); dr.pop()
        # copy a value only
        nd[2] = y[2]
    else:
        y = nd

    # delete the node (y) with at most one non-leaf child
    for x in st:
        x[3] -= 1

    c = y[0] if y[0] else y[1]
    if st:
        st[-1][dr[-1]] = c

    if y[4]:
        # if nd is red
        return st[0] if st else c

    if c and c[4]:
        # if nd is black and child is red
        c[4] = 0
        return st[0] if st else c

    # nd and child is black: delete in each cases

    n = c
    while st:
        p = st.pop(); dp = dr.pop()
        # n's sibling
        s = p[dp ^ 1]
        #assert s
        sc = s[dp]; sd = s[dp ^ 1]

        if s[4]:
            #assert sc and sd
            p[4] = 1; s[4] = 0
            if st:
                st[-1][dr[-1]] = rotate(p, dp ^ 1)
            else:
                rotate(p, dp ^ 1)
            st.append(s); dr.append(dp)
            #assert p[dp ^ 1] is sc
            s = sc
            sc = s[dp]; sd = s[dp ^ 1]
            break

        if p[4] or (sc and sc[4]) or (sd and sd[4]):
            break

        s[4] = 1
        n = p
    else:
        return n

    if p[4] and not s[4] and not (sc and sc[4]) and not (sd and sd[4]):
        p[4] = 0; s[4] = 1
        return st[0] if st else p

    sc = s[dp]; sd = s[dp ^ 1]
    if not s[4] and (sc and sc[4]) and not (sd and sd[4]):
        sc[4] = 0; s[4] = 1
        sd = s
        s = p[dp ^ 1] = rotate(s, dp)

    #assert not s[4] and (sd and sd[4])
    s[4] = p[4]
    p[4] = 0
    sd[4] = 0

    if st:
        st[-1][dr[-1]] = rotate(p, dp ^ 1)
    else:
        return rotate(p, dp ^ 1)

    return st[0] if st else n

# delete a node with key = val from a tree t
def delete(t, val):
    x = t

    st = []; dr = []
    while x and val != x[2]:
        d = (x[2] < val)
        st.append(x); dr.append(d)
        x = x[d]
    if not x:
        return t
    # x is the node with key = val
    return __delete(st, dr, x)

File: cpp/binary_search_tree/red-black-tree.cpp
template<typename T>
class RedBlackTree {
  enum NODE_COLOR { BLACK, RED };
  struct Node {
    Node *left, *right, *prt;
    T key;
    int color;
    int size;

    Node(T x) {
      left = right = prt = nullptr;
      color = RED;
      key = x;
      size = 1;
    }

    inline bool is_left(Node *node) const {
      return this->left == node;
    }

    inline void assign(Node *node) {
      this->key = node->key;
    }

    inline Node* rotate_left() {
      Node *r = this->right, *m = r->left, *p = this->prt;
      if((r->prt = p)) {
        if(p->left == this) p->left = r;
        else p->right = r;
      }
      if((this->right = m)) m->prt = this;
      r->left = this; this->prt = r;

      int sz = this->size;
      this->size += (m ? m->size : 0) - r->size;
      r->size = sz;

      return r;
    }

    inline Node* rotate_right() {
      Node *l = this->left, *m = l->right, *p = this->prt;
      if((l->prt = p)) {
        if(p->left == this) p->left = l;
        else p->right = l;
      }
      if((this->left = m)) m->prt = this;
      l->right = this; this->prt = l;

      int sz = this->size;
      this->size += (m ? m->size : 0) - l->size;
      l->size = sz;

      return l;
    }

    inline Node* get_sib(Node *node) {
      return (this->is_left(node) ? this->right : this->left);
    }
  };

  Node *root;

  inline Node* find_node(Node *node, T x) {
    if(node == nullptr) return nullptr;
    while(node->key != x) {
      if(x < node->key) {
        if(!node->left) break;
        node = node->left;
      } else {
        if(!node->right) break;
        node = node->right;
      }
    }
    return node;
  }

  inline static bool is_red(Node *node) {
    return node && node->color == RED;
  }

  inline static bool is_black(Node *node) {
    return !node || node->color == BLACK;
  }

  inline void remove_node(Node *node) {
    Node *prt = node->prt;

    if(node->left && node->right) {
      Node *c_node = find_node(node->right, node->key);
      node->assign(c_node);
      node = c_node; prt = c_node->prt;
    }
    Node *dnode = node;

    Node *n_node = (node->left ? node->left : node->right);
    if(prt) {
      if(node->key < prt->key) {
        prt->left = n_node;
      } else {
        prt->right = n_node;
      }
    }
    if((node = n_node)) node->prt = prt;

    Node *cur = prt;
    while(cur) {
      --cur->size;
      cur = cur->prt;
    }

    if(is_red(dnode)) return;
    if(is_red(node)) {
      node->color = BLACK;
      if(!prt) this->root = node;
      return;
    }

    while(prt) {
      Node *sib = prt->get_sib(node);

      if(is_red(sib)) {
        if(prt->is_left(node)) {
          prt->rotate_left();
        } else {
          prt->rotate_right();
        }

        prt->color = RED;
        sib->color = BLACK;

        sib = prt->get_sib(node);
      }

      if(is_red(prt) || is_red(sib) || is_red(sib->left) || is_red(sib->right)) {
        if(is_red(prt) && is_black(sib) && is_black(sib->left) && is_black(sib->right)) {
          sib->color = RED;
          prt->color = BLACK;
          break;
        }

        if(is_black(sib)) {
          if(prt->is_left(node) && is_black(sib->right) && is_red(sib->left)) {
            Node *r = sib->rotate_right();
            r->color = RED;
            r->right->color = BLACK;
          } else if(!prt->is_left(node) && is_black(sib->left) && is_red(sib->right)) {
            Node *r = sib->rotate_left();
            r->color = RED;
            r->left->color = BLACK;
          }
          sib = prt->get_sib(node);
        }

        if(prt->is_left(node)) {
          prt->rotate_left();

          sib->color = prt->color;
          prt->color = sib->right->color = BLACK;
        } else {
          prt->rotate_right();

          sib->color = prt->color;
          prt->color = sib->left->color = BLACK;
        }
        break;
      }

      sib->color = RED;

      node = prt; prt = node->prt;
    }

    Node *n = (prt ? prt : node);
    if(n) {
      while(n->prt) n = n->prt;
    }
    this->root = n;
  }

public:
  RedBlackTree() {
    this->root = nullptr;
  }

  inline bool find(T x) {
    if(!this->root) return false;

    Node *node = find_node(this->root, x);
    return (node->key == x);
  }

  inline T at(int k) {
    if(!this->root) {
      return 0;
    }
    // assert(0 <= k < size);

    Node *node = this->root;
    ++k;
    while(1) {
      int l_size = (node->left ? node->left->size : 0) + 1;
      if(l_size == k) break;

      if(k < l_size) {
        node = node->left;
      } else {
        node = node->right;
        k -= l_size;
      }
    }
    return node->key;
  }

  inline bool insert(T x) {
    if(!this->root) {
      Node *new_node = new Node(x);
      this->root = new_node;
      new_node->color = BLACK;
      return true;
    }

    Node *node = find_node(this->root, x);
    if(node->key == x) return false;

    Node *new_node = new Node(x);
    if(x < node->key) {
      node->left = new_node;
    } else {
      node->right = new_node;
    }
    new_node->prt = node;

    while(node) {
      ++node->size;
      node = node->prt;
    }

    node = new_node;
    while(is_red(node->prt)) {
      Node *prt = node->prt;

      Node *gprt = prt->prt;
      Node *u = gprt->get_sib(prt);
      if(is_black(u)) {
        if(gprt->is_left(prt) && !prt->is_left(node)) {
          prt = prt->rotate_left();
          node = prt->left;
        } else if(!gprt->is_left(prt) && prt->is_left(node)) {
          prt = prt->rotate_right();
          node = prt->right;
        }

        if(prt->is_left(node)) {
          gprt->rotate_right();
        } else {
          gprt->rotate_left();
        }

        prt->color = BLACK;
        gprt->color = RED;

        break;
      }

      prt->color = u->color = BLACK;
      gprt->color = RED;

      node = gprt;
    }
    if(!node->prt) node->color = BLACK;

    while(node->prt) node = node->prt;
    this->root = node;

    return true;
  }

  inline bool remove(T x) {
    if(!this->root) {
      return false;
    }

    Node *node = find_node(this->root, x);
    if(node->key != x) {
      return false;
    }

    remove_node(node);
    return true;
  }

  int size() {
    return (this->root ? this->root->size : 0);
  }
};
